staging/src/k8s.io/apiserver/pkg/registry/rest/create.go

```diff
@@ -145,21 +147,31 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 
 // CheckGeneratedNameError checks whether an error that occurred creating a resource is due
 // to generation being unable to pick a valid name.
-func CheckGeneratedNameError(strategy RESTCreateStrategy, err error, obj runtime.Object) error {
+func CheckGeneratedNameError(ctx context.Context, strategy RESTCreateStrategy, err error, obj runtime.Object) error {
 	if !errors.IsAlreadyExists(err) {
 		return err
 	}
 
-	objectMeta, kind, kerr := objectMetaAndKind(strategy, obj)
+	objectMeta, gvk, kerr := objectMetaAndKind(strategy, obj)
 	if kerr != nil {
 		return kerr
 	}
 
 	if len(objectMeta.GetGenerateName()) == 0 {
```

Comment:
My only remaining thought is, can we make the message less misleading if the user sets both?

What if we do something like this:

```
if gn, n := objectMeta.GetGenerateName(), objectMeta.GetName(); len(gn) == 0 || !strings.HasPrefix(n, gn) {
```

It's not perfect, but this should at least detect cases where user has set obviously conflicting generateName & name?

We can do it in a followup if this is making it too complicated.

Reply:
This is a great idea, although we might need more signal that the above comparison, because the `generateName` can still be a prefix of `name`. In most cases, users setting both have probably gotten current objects through kubectl or other clients, and then tried to recreate them.

Reply:
Yeah, people in that situation will have to ask themselves, "hm why does the server choose the same name every time??"

I'd fix that too but I think it requires more plumbing and I've already made enough requests on this PR :)

Reply:
Yeah, let's merge this as-is and iterate on it as we go?